### DISCUSSION TITLE
How to add a new statement in the immediate block ?

### DISCUSSION THREAD

**Discussion by yogeshgadge** [babel/babel#15827] [Q&A] [answered]
Gist with example and partial transformer- 

https://astexplorer.net/#/gist/77311bc11f39d25e656b7040801445b9/a24e52765de1a985e13872ddccf7a1c1bff6de5c

I would like to catch "await in arguments" and replace by local statements in the immediate block.
(Background - trying to use some aws classes directly in nginx njs and  it faiils at something like this `toHex(foo, await digest(), await bar)` saying "SyntaxError: await in arguments not supported in  SyntaxError: await in arguments not supported in.."

The full example and partial transformer is in the gist but here is the example:-

```JavaScript
async function greet(name) {
    console.log(await digest());
    const foo = 'foo';
    const bar = 'bar';
    const b = toHex(foo, await digest(), await bar);
    return b
}

function toHex(a, b, c) {
    return true;
}
async function digest() {
    return 'v';
}
```

to something like this:-

```JavaScript
async function greet(name) {
    console.log(await digest());
    const foo = 'foo';
    const bar = 'bar';
    const someVar1 = await digest();
    const someVar2 =  await bar;
    const b = toHex(foo, someVar1, someVar2);
    return b
}

function toHex(a, b, c) {
    return true;
}
async function digest() {
    return 'v';
}
```

I believe the following condition provides me with correct arguments. But How do I go about replacing the arguments with some new constants assignment statements which I also want to inject before.

```
module.exports = function (babel) {
  const { types: t, template } = babel;
  
  return {
    name: "change-gatsby-function", // not required
    visitor: {
      CallExpression(path) {
        const node = path.node;
        const awaitArgs = node.arguments.filter(a => a.type === 'AwaitExpression');
        if (awaitArgs.length <= 0) {
          return;
        }
        //if (path.parent.callee.name === "useMemo") {
        //  path.replaceWith(
        //    functionExpression
        //  );
        //}
      }
    }
  };
};

``` 

I am new to babel plugins development after years of using other's work (thank you) but I am thinking that i could some how traverse the `path.parentPath` and check for some `BlockStatement` and then inject some snippet?

**Comment by nicolo-ribaudo** [comment, marked as answer]
I would visit `Statement` instead, and from there have a nested traversal to find all the `await`s in call arguments: https://astexplorer.net/#/gist/18764696ca1f513e1ddacd69cb10bf78/53829615938a05b0771e5dcd463b9bc8d5fe229a

EDIT: This also works with nested calls: https://astexplorer.net/#/gist/18764696ca1f513e1ddacd69cb10bf78/c81c58c6b09e8154c0b0cb327a848b3d556bdfaa

**Comment by yogeshgadge** [comment]
Thank you - the solution in not quite working in my local:-
It is missing the variable declaration - wondering any clue. It is doing correctly for the `console.log(await digest());` as it is creating `_digest` const declaration

```
async function greet(name) {
  const _digest = await digest();
  console.log(_digest);
  const foo = 'foo';
  const bar = 'bar';
  await foo;
    await digest();
    await bar;
  const b = toHex();
  return b;
}
```

**Reply by nicolo-ribaudo** [reply]
That is wierd, could you share the full plugin code if you modified it? 🤔

**Comment by yogeshgadge** [comment]
Nevermind the issue is when running through roll-up babel plugin. The
plugin actually works when running without roll-up when tried cli.
Investigating further from roll-up side.

#close
